### PR TITLE
HIVE-25801: Custom queue setting is not honoured by Query based compaction StatsUpdater

### PR DESCRIPTION
**Pull request by veghlaci05**
### What changes were proposed in this pull request?

This PR fixes the issue that the statistics gathering for Query based compactions are not honouring the custom queue settings. It also deprecates the hive.mr.compactor.gather.stats setting which is honoured only by the MR based compactor, and introduces the hive.compactor.gather.stats setting, which is honoured by both MR and Query based compactors. The values of the two settings are combined with logical OR in case of MR based compaction. The stats gathering also changed since there is no need for column based statistics gathering in case of compaction, the basic statistics should be enough.

### Why are the changes needed?

hive.compactor.job.queue config limits resources available for compaction, so users can limit the effects of compaction on the cluster. However this setting does not affect stats collection which uses Driver.
HIVE-25595 is addressing the above issue for MR-based compaction. We need to incorporate the same thing for the Query-based compaction.

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
It was tested using the already existing automated tests.

### CODE REVIEW

File: ql/src/java/org/apache/hadoop/hive/ql/exec/tez/DagUtils.java [hunk 1]
```diff
@@ -1299,6 +1299,10 @@ public static FileStatus validateTargetDir(Path path, Configuration conf) throws
   // the api that finds the jar being used by this class on disk
   public String getExecJarPathLocal(Configuration configuration) throws URISyntaxException {
     // returns the location on disc of the jar of this class.
+    String jar = configuration.get(ConfVars.HIVEJAR.varname);
+    if (!StringUtils.isBlank(jar)) {
+      return jar;
```
Comment (deniskuzZ) [inline ql/src/java/org/apache/hadoop/hive/ql/exec/tez/DagUtils.java:1304]: why is this change needed?

File: ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/QueryCompactor.java [hunk 2]
```diff
@@ -137,6 +140,10 @@ void runCompactionQueries(HiveConf conf, String tmpTableName, StorageDescriptor
         DriverUtils.runOnDriver(conf, user, sessionState, query, writeIds, compactorTxnId);
       }
       commitCompaction(storageDescriptor.getLocation(), tmpTableName, conf, writeIds, compactorTxnId);
+
+      if (su != null) {
```
Comment (deniskuzZ) [inline ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/QueryCompactor.java:144]: why not call it from Worker, on compaction txn commit?
````
msc.commitTxn(txnId);
status = TxnStatus.COMMITTED;
su.gatherStats()
````

File: ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/Worker.java [hunk 3]
```diff
@@ -506,10 +478,23 @@ protected Boolean findNextCompactionAndExecute(boolean computeStats) {
         task (currently we're using Tez split grouping).
         */
         QueryCompactor queryCompactor = QueryCompactorFactory.getQueryCompactor(t, conf, ci);
+        boolean generic_stats = conf.getBoolVar(HiveConf.ConfVars.HIVE_COMPACTOR_GATHER_STATS);
```
Comment (deniskuzZ) [inline ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/Worker.java:481]: Why did you move the config evaluation inside of the compaction loop? 
Please use camelCase naming convention.

File: ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/Worker.java [hunk 4]
```diff
@@ -506,10 +478,23 @@ protected Boolean findNextCompactionAndExecute(boolean computeStats) {
...
+        boolean mr_stats = conf.getBoolVar(HiveConf.ConfVars.HIVE_MR_COMPACTOR_GATHER_STATS);
+        boolean computeStats = (queryCompactor == null && (mr_stats || generic_stats)) || generic_stats;
+
+        LOG.info("Starting " + ci.type.toString() + " compaction for " + ci.getFullPartitionName() + " in " +
+                compactionTxn + " with compute stats set to " + computeStats);
+
+        StatsUpdater su = null;
+        if (computeStats) {
+          su = StatsUpdater.init(ci, conf, runJobAsSelf(ci.runAs) ? ci.runAs : t.getOwner(),
```
Comment (deniskuzZ) [inline ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/Worker.java:490]: Q. can we refactor StatsUpdater to a service utility that just submits the compute stats query and not generate a separate instance per each compaction candidate?

File: ql/src/java/org/apache/hadoop/hive/ql/exec/tez/DagUtils.java [hunk 5]
```diff
@@ -1297,18 +1297,20 @@ public static FileStatus validateTargetDir(Path path, Configuration conf) throws
...
-    URI uri = DagUtils.class.getProtectionDomain().getCodeSource().getLocation().toURI();
-    if (configuration.getBoolean(ConfVars.HIVE_IN_TEST_IDE.varname, false)) {
-      if (new File(uri.getPath()).isDirectory()) {
-        // IDE support for running tez jobs
-        uri = createEmptyArchive();
+  public String getExecJarPathLocal(Configuration configuration) {
+    try {
+      // returns the location on disc of the jar of this class.
+      String uri = DagUtils.class.getProtectionDomain().getCodeSource().getLocation().toURI().toString();
+      if (uri.endsWith(".jar")) {
```
Comment (deniskuzZ) [inline ql/src/java/org/apache/hadoop/hive/ql/exec/tez/DagUtils.java:1304]: can we get NPE here?

File: ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/QueryCompactor.java [hunk 6]
```diff
@@ -56,6 +56,7 @@
    * @param storageDescriptor this is the resolved storage descriptor
    * @param writeIds valid write IDs used to filter rows while they're being read for compaction
    * @param compactionInfo provides info about the type of compaction
+   * @param dir provides ACID directory layout information
```
Comment (deniskuzZ) [inline ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/QueryCompactor.java:59]: redundant

File: ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/Worker.java [hunk 7]
```diff
@@ -87,17 +87,18 @@
   private String workerName;
 
   // TODO: this doesn't check if compaction is already running (even though Initiator does but we
-  // don't go through Initiator for user initiated compactions)
+  //  ,mdon't go through Initiator for user initiated compactions)
```
Comment (deniskuzZ) [inline ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/Worker.java:90]: nit, typo

File: ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/Worker.java [hunk 8]
```diff
@@ -505,19 +472,29 @@ protected Boolean findNextCompactionAndExecute(boolean computeStats) {
         task (currently we're using Tez split grouping).
         */
         QueryCompactor queryCompactor = QueryCompactorFactory.getQueryCompactor(t, conf, ci);
+        boolean computeStats = (queryCompactor == null && (collectMrStats || collectGenericStats)) || collectGenericStats;
```
Comment (deniskuzZ) [inline ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/Worker.java:475]: you can simplify
````
computeStats = (queryCompactor == null && collectMrStats) || collectGenericStats;
````

File: itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/txn/compactor/TestCrudCompactorOnTez.java [hunk 9]
```diff
@@ -68,10 +59,16 @@
 import org.junit.Assert;
 import org.junit.Test;
 
-import static org.apache.hadoop.hive.ql.txn.compactor.TestCompactor.executeStatementOnDriver;
+import java.io.IOException;
```
Comment (deniskuzZ) [inline itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/txn/compactor/TestCrudCompactorOnTez.java:62]: why is the import reorder?

Comment (veghlaci05) [reply itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/txn/compactor/TestCrudCompactorOnTez.java:62]: There were some unused imports and I used Intellij Idea;s 'organize imports' command which orders and groups the imports as well on top of removing the unused ones.

File: ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/Worker.java (hunk 3 above)
Comment (veghlaci05) [reply ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/Worker.java]: I moved it back

File: itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/txn/compactor/TestCrudCompactorOnTez.java (hunk 9 above)
Comment (veghlaci05) [reply itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/txn/compactor/TestCrudCompactorOnTez.java]: Reverted the reorder, only removed the unused ones

File: ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/Worker.java (hunk 8 above)
Comment (veghlaci05) [reply ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/Worker.java]: Applied

File: ql/src/java/org/apache/hadoop/hive/ql/exec/tez/DagUtils.java (hunk 5 above)
Comment (veghlaci05) [reply ql/src/java/org/apache/hadoop/hive/ql/exec/tez/DagUtils.java:1304]: Should not, but modified the catch block to Exception class, just to make sure no exception is thrown from this method.

File: itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/txn/compactor/TestCompactor.java [hunk 10]
```diff
@@ -335,153 +330,144 @@ public void schemaEvolutionAddColDynamicPartitioningUpdate() throws Exception {
...
+    CompactionInfo ci = new CompactionInfo(dbName, tblName, "bkt=0", CompactionType.MAJOR);
+    Worker.StatsUpdater.gatherStats(ci, conf,
+            System.getProperty("user.name"), CompactorUtil.getCompactorJobQueueName(conf, ci, table));
+    ci = new CompactionInfo(dbName, tblName, "bkt=1", CompactionType.MAJOR);
+    Worker.StatsUpdater.gatherStats(ci, conf,
+            System.getProperty("user.name"), CompactorUtil.getCompactorJobQueueName(conf, ci, table));
+
+    //Check basic stats are collected
+    org.apache.hadoop.hive.ql.metadata.Table hiveTable = Hive.get().getTable(tblName);
+    Map<String, String> parameters = Hive.get().getPartitions(hiveTable).get(0).getParameters();
```
Comment (deniskuzZ) [inline itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/txn/compactor/TestCompactor.java:377]: should we check both partitions as you are loading data into both?

File: itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/txn/compactor/TestCompactor.java [hunk 11]
```diff
@@ -90,7 +67,25 @@
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
-import com.google.common.collect.Lists;
+import java.io.File;
```
Comment (deniskuzZ) [inline itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/txn/compactor/TestCompactor.java:70]: imports structure is not restored

File: ql/src/test/org/apache/hadoop/hive/ql/TestTxnNoBuckets.java [hunk 12]
```diff
@@ -833,10 +833,6 @@ public void testCompactStatsGather() throws Exception {
     Assert.assertEquals("Unexpected number of compactions in history", 1, resp.getCompactsSize());
     Assert.assertEquals("Unexpected 0 compaction state", TxnStore.CLEANING_RESPONSE, resp.getCompacts().get(0).getState());
     Assert.assertTrue(resp.getCompacts().get(0).getHadoopJobId().startsWith("job_local"));
-
-    //now check that stats were updated
-    map = hms.getPartitionColumnStatistics("default","T", partNames, colNames, Constants.HIVE_ENGINE);
```
Comment (deniskuzZ) [inline ql/src/test/org/apache/hadoop/hive/ql/TestTxnNoBuckets.java:838]: should we check basic statistics here instead on colStat?

File: common/src/java/org/apache/hadoop/hive/conf/HiveConf.java [hunk 13]
```diff
@@ -3257,11 +3257,23 @@ private static void populateLlapDaemonVarsSet(Set<String> llapDaemonVarsSetLocal
...
+     * Use the {@link HiveConf.ConfVars#HIVE_COMPACTOR_GATHER_STATS}
+     * config instead which is honoured by both the MR and Query based compaction.
+     */
+    @Deprecated
+    HIVE_MR_COMPACTOR_GATHER_STATS("hive.mr.compactor.gather.stats", false, "If set to true MAJOR compaction " +
         "will gather stats if there are stats already associated with the table/partition.\n" +
         "Turn this off to save some resources and the stats are not used anyway.\n" +
         "Works only for MR based compaction, CRUD based compaction uses hive.stats.autogather."),
 
+    HIVE_COMPACTOR_GATHER_STATS("hive.compactor.gather.stats", true, "If set to true MAJOR compaction " +
```
Comment (deniskuzZ) [inline common/src/java/org/apache/hadoop/hive/conf/HiveConf.java:3271]: Do we re-calculate stats only in case of MAJOR compaction?

Comment (veghlaci05) [reply common/src/java/org/apache/hadoop/hive/conf/HiveConf.java:3271]: Yes, this is the current behavior

File: ql/src/test/org/apache/hadoop/hive/ql/TestTxnNoBuckets.java (hunk 12 above)
Comment (veghlaci05) [reply ql/src/test/org/apache/hadoop/hive/ql/TestTxnNoBuckets.java:838]: Modified the test to check basic stats only
